### PR TITLE
Single Routing Entrypoint

### PR DESCRIPTION
**Pull request by jeyj0**
This PR introduces an optional `router` function for `FrontController`s, which allows handling the routing for that FrontController only in that function, or extending the `defaultRouter`, which is the previous behavior. Additionally, the function allows using `putContextRouter`, a new function that is used to modify the `ControllerContext` during routing, ~*before*~ ~*after*~ *before* `initContext` is called.

The default implementation for `router` is `router = defaultRouter`, which is 100% backwards-compatible.

An example of why this is useful and how it is used is this router implementation:

```haskell
instance FrontController WebApplication where
    ... -- the previous code assigning controllers is still required

    router additionalControllers = do
        mlocale <- optional do
            string "/"
            string "en"

        let router = defaultRouter additionalControllers

        case mlocale of
            Just locale -> case locale of
                "en" -> putContextRouter EN_UK router
                _ -> IHP.RouterPrelude.error "This should be unreachable"
            Nothing -> router
```

This handles all routes as before but also allows an optional `/en` to be added in front of all routes. In case it is added, the `ControllerContext` is modified to contain a locale (`EN_UK`), which allows the application to show in English (instead of the default German for this application).

~The `additionalControllers` are the `controllers` of the `FrontController`, and the other controllers added by IHP by default (for example for DataSync).~
The `additionalControllers` are the controllers added by IHP by default (for example for DataSync). The controllers of the `FrontController` itself are accessible via the typeclass' `controllers` function.

Since `controllers` returns a list of `RouteParser`s, it is possible to nest these custom routers using `mountFrontController` as before. This means it is possible to override all routing by implementing `router` for the `RootApplication`.

---

Since implementing the `router` function can change how routes are handled, it can also break using `pathTo`, `urlTo`, or implicit `pathTo` in HSX. In the future, `pathTo` should be made `ControllerContext`-aware so those methods can be compatible with the `router`. In the meantime, I suggest not adding this feature to the documentation yet, to avoid unnecessary confusion.

As an example, in the above application, none of `pathTo`, `urlTo`, or implicit `pathTo` in HSX would ever result in an `/en` being added to the url. By making `pathTo` `ControllerContext`-aware, it could read the current locale from context, and prepend `/en` to the path if the locale is `EN_UK`. As a workaround for this application, a `localizedPathTo` and `localizedUrlTo` will be added, which does this, but care has to be taken to use these functions instead of `pathTo` and `urlTo`, and implicit `pathTo` cannot be used.

---

**Edit:** updated the PR text to match actual behavior that I wrongly remembered.

### CODE REVIEW

File: IHP/RouterSupport.hs [hunk 1]
```diff
@@ -68,9 +69,81 @@ import qualified IHP.WebSocket as WS
...
+        => [RouteParser] -> RouteParser
+    router = defaultRouter
+
+    defaultRouter
+        :: (?applicationContext :: ApplicationContext, ?application :: application, ?context :: RequestContext)
+        => [RouteParser] -> RouteParser
+    defaultRouter additionalControllers = do
+        let allControllers = controllers <> additionalControllers
+        ioResponseReceived <- choice $ map (\r -> r <* endOfInput) allControllers
+        pure ioResponseReceived
```
Comment (mpscholten) [inline IHP/RouterSupport.hs:146]: It seems like this could be moved out of the type class into a standalone function

File: IHP/RouterSupport.hs [hunk 2]
```diff
@@ -68,9 +69,81 @@ import qualified IHP.WebSocket as WS
 import GHC.TypeLits as T
 import IHP.Controller.Context
 import IHP.Controller.Param
+import qualified Data.TMap as TMap
+import qualified IHP.ApplicationContext as ApplicationContext
+
+putContextRouter :: forall value. (Typeable value) => value -> RouteParser -> RouteParser
```
Comment (mpscholten) [inline IHP/RouterSupport.hs:75]: Why can't we use the existing controller context also during routing?

Right now the controller context is created when the action is called. But it seems there's nothing preventing us from creating the controller context before routing.

Then we could just use `putContext` eveywhere, which would simplify the public API of IHP.

File: IHP/ControllerSupport.hs [hunk 3]
```diff
@@ -99,13 +105,23 @@ runActionWithNewContext controller = do
 
     try (initContext @application) >>= \case
         Left (exception :: SomeException) -> do
-            case fromException exception of
+            pure $ Left $ case fromException exception of
```
Comment (mpscholten) [inline IHP/ControllerSupport.hs:108]: Why is this change needed?

File: IHP/RouterSupport.hs (hunk 2 above)
Comment (jeyj0) [reply IHP/RouterSupport.hs:75]: `putContext` is an `IO` function. So parsing would require constant lifts, which to me seemed less nice than having to use a special `putContextRouter` function for parsing.

Comment (mpscholten) [reply IHP/RouterSupport.hs:75]: Ok this makes sense 👍

File: IHP/ControllerSupport.hs (hunk 3 above)
Comment (jeyj0) [reply IHP/ControllerSupport.hs:108]: I think here, the diff is mostly confusing. What I've done is taken the logic from `runActionWithNewContext` and split it into a `newContextForAction` and `runActionWithContext`, then implementing the old `runActionWithNewContext` in terms of those two new functions. So the old function has the same signature and does the same thing. I just needed separate access to the two steps within `RouterSupport`, because I needed a way to create a `ControllerContext` with `initContext` and all other defaults in mind, apply changes to it (from the router), and then run an action with the updated `ControllerContext`.

While writing this, I am realizing that my explanation in the PR is wrong though: first, `initContext` is executed, then the changes by the `router` are applied, not the other way around. I'll have to think about which order makes more sense.

**Comment by jeyj0** [comment]
As written above in a comment, currently, the order of creating the `ControllerContext` before the action is called is:
1. IHP defaults
2. `initContext`
3. `router` changes are applied

I think the more useful order is:
1. IHP defaults
2. `router` changes are applied
3. `initContext`

This way, the `router`'s changes are available in `initContext`, where any generic logic can be executed due to the current context state. For example, if from routing it is unclear which locale should be used for the app, `initContext` can use the `Accept-Language` header to detect it, and possibly redirect to a better route.

The current order does not have any benefit, as the `router` has far less flexibility in what it can do. For example, the above use case would require a change in all actions to work properly.

I will update the order of creating the `ControllerContext`.

File: IHP/ControllerSupport.hs [hunk 4]
```diff
@@ -87,25 +89,61 @@ runAction controller = do
 
     doRunAction `catches` [ Handler handleResponseException, Handler (\exception -> ErrorController.displayException exception controller "")]
 
-{-# INLINE runActionWithNewContext #-}
-runActionWithNewContext :: forall application controller. (Controller controller, ?applicationContext :: ApplicationContext, ?context :: RequestContext, InitControllerContext application, ?application :: application, Typeable application, Typeable controller) => controller -> IO ResponseReceived
-runActionWithNewContext controller = do
+{-# INLINE runActionWithContext #-}
+runActionWithContext :: forall application controller. (Controller controller, ?modelContext :: ModelContext, ?applicationContext :: ApplicationContext, ?requestContext :: RequestContext) => ControllerContext -> controller -> IO ResponseReceived
+runActionWithContext controllerContext controller = let ?context = controllerContext in runAction controller
```
Comment (mpscholten) [inline IHP/ControllerSupport.hs:94]: This function IMO is the same as `runAction`. Can't we then just remove this function entirely?

File: IHP/ControllerSupport.hs [hunk 5]
```diff
@@ -87,25 +89,61 @@ runAction controller = do
...
+
+applyContextSetter :: (TypeMap.TMap -> TypeMap.TMap) -> ControllerContext -> IO ControllerContext
+applyContextSetter setter ctx@ControllerContext { customFieldsRef } = do
+    modifyIORef customFieldsRef (applySetter setter)
```
Comment (mpscholten) [inline IHP/ControllerSupport.hs:98]: The strict version `modifyIORef'` might be better here. Otherwise `applyContextSetter` might not catch `error` calls that could be created in the router system until the context is actually used.

File: IHP/ControllerSupport.hs (hunk 3 above)
Comment (mpscholten) [reply IHP/ControllerSupport.hs:130]: > While writing this, I am realizing that my explanation in the PR is wrong though: first, initContext is executed, then the changes by the router are applied, not the other way around. I'll have to think about which order makes more sense.

As a user I'd naively expect the router context to come first and be accessible from the `initContext`

Comment (jeyj0) [reply IHP/ControllerSupport.hs:130]: I have already changed it to this behavior below :+1:

File: IHP/ControllerSupport.hs (hunk 4 above)
Comment (jeyj0) [reply IHP/ControllerSupport.hs]: Yes, you are absolutely correct. Probably a holdover from some changes I made. Fixed!

File: IHP/ControllerSupport.hs (hunk 5 above)
Comment (jeyj0) [reply IHP/ControllerSupport.hs]: Agreed. Fixed.

**Review by mpscholten — APPROVED** [review]

**Comment by mpscholten** [comment]
Sorry, I removed it because I thought not many people were using it. Additionally it caused a lot of complexity in the routing.

Here is an example of how this can be done using the request vault (a feature of wai, the underlying http server library IHP uses, it's like a simple key value object, see https://hackage.haskell.org/package/vault):

Let's assume we have a function to enable i18n in routing:
```haskell
instance FrontController WebApplication where
    controllers =
        [ startPage WelcomeAction
        -- ...
        ]

    router additionalControllers = do
        mlocale <- optional do
            string "/"
            string "en" -- <|> string "de"

        let router = defaultRouter additionalControllers
        case mlocale of
            Just locale -> case locale of
                "en" -> putContextRouter EN_UK router
                -- "de" -> putContextRouter DE_DE router
                _ -> IHP.RouterPrelude.error "This should be unreachable"
            Nothing -> putContextRouter defaultLocale router
        defaultRouter additionalControllers
```

This function allows urls such as `/en/...` and `/de/...` to be routed and also the language will be stored in the context via `putContextRouter DE_DE`.

The new way is to write a middleware that stores it into the request vault:

```haskell
import qualified Data.Vault.Lazy as Vault
import System.IO.Unsafe (unsafePerformIO)
import Network.Wai

localeVaultKey :: Vault.Key Locale
localeVaultKey = unsafePerformIO Vault.newKey
{-# NOINLINE localeVaultKey #-}

localeMiddleware :: Middleware
localeMiddleware app request respond =
    let
        locale = localeFromPathInfo request.pathInfo
        vault' = Vault.insert localeVaultKey locale request.vault
        request' = request { vault = vault' }
    in
        app request' respond

localeFromPathInfo :: [Text] -> Locale
localeFromPathInfo pathInfo =
    case pathInfo of
        ("en":_) -> EN_UK
        otherwise -> DE_DE
```

Once enabled, the locale can be access from the vault instead of the context directly:
```
-- OLD:
-- currentLocale :: _ => Locale
-- currentLocale = fromMaybe defaultLocale $ maybeFromFrozenContext @Locale
-- NEW:
currentLocale :: (?context :: ControllerContext) => Locale
currentLocale =
    Vault.lookup localeVaultKey request.vault
    |> fromMaybe (error "no locale set")
```

The middleware also needs to be activated in the `Config.hs`:

```

config :: ConfigBuilder
config = do
    option (CustomMiddleware localeMiddleware)
```

Now `putContextRouter` is not needed anymore.

**Comment by stephenbenedict** [comment]
@mpscholten Wow! Thank you so much for the comprehensive response. I will try that.

**Comment by stephenbenedict** [comment]
Hm, I must be missing something. Everything compiles fine, but routing with the locale prefix (e.g. /ja ) is not working (get 404).

```haskell
-- Config.hs
import Application.Helper.Localization

config :: ConfigBuilder
config = do
    option (CustomMiddleware localeMiddleware)
```

```haskell
-- Localization.hs
{-# LANGUAGE DeriveGeneric #-}
module Application.Helper.Localization
    ( Localizable(..)
    , l10n
    , l10nText
    , defaultLocale
    , currentLocale
    , localeMiddleware
    ) where

import IHP.ViewPrelude
import Generated.Types
import IHP.ControllerPrelude
import qualified Data.Vault.Lazy as Vault
import System.IO.Unsafe (unsafePerformIO)
import Network.Wai

localeVaultKey :: Vault.Key Locale
localeVaultKey = unsafePerformIO Vault.newKey
{-# NOINLINE localeVaultKey #-}

localeMiddleware :: Middleware
localeMiddleware app request respond =
    let
        locale = localeFromPathInfo request.pathInfo
        vault' = Vault.insert localeVaultKey locale request.vault
        request' = request { vault = vault' }
    in
        app request' respond

localeFromPathInfo :: [Text] -> Locale
localeFromPathInfo pathInfo =
    case pathInfo of
        ("ja":_) -> JaJp
        otherwise -> EnUs

currentLocale :: (?context :: ControllerContext) => Locale
currentLocale =
    Vault.lookup localeVaultKey request.vault
    |> fromMaybe (error "no locale set")

data Localizable
    = Heading
    | Introduction
    | Body

l10nText :: _ => Localizable -> Text
l10nText l = l10nText' l currentLocale
```

**Comment by stephenbenedict** [comment]
Figured it out. Forgot that I needed to keep this in Web.FrontController:

```haskell
router additionalControllers = do
        mlocale <- optional do
            string "/"
            string "ja"
        defaultRouter additionalControllers  
```
